src/controllers/resize.ts

```diff
@@ -0,0 +1,62 @@
+import { Context } from 'koa'
+import { Controller, Ctx, Get, Params, Query } from 'amala'
+import { IPFS_TIMEOUT } from '../constants'
+import { Readable } from 'stream'
+import { badRequest, internal, notFound } from '@hapi/boom'
+import { getCachedResize, setCachedResize } from '../helpers/resizedImages'
+import { getContentType } from '../helpers/contentTypes'
+import ipfs, { getDataFromIPFS } from '../helpers/ipfs'
+import sharp from 'sharp'
+
+@Controller('/resize')
+export default class ResizeController {
+  @Get('/image/:cid')
+  async image(
+    @Ctx() ctx: Context,
+    @Params('cid') cid: string,
+    @Query('width') width?: number,
+    @Query('height') height?: number
+  ) {
+    try {
+      const contentType = await getContentType(cid)
+      if (contentType) ctx.set('Content-Type', contentType)
+      if (contentType && !contentType.startsWith('image/'))
+        return ctx.throw(400, badRequest('Not an image'))
+
+      // If no width or height is provided, return the original image
+      // If the image is a gif, return the original image
+      if ((!width && !height) || (contentType && contentType.endsWith('gif')))
+        return Readable.from(ipfs.cat(cid, { timeout: IPFS_TIMEOUT }))
+
+      // Check for cache hit; if found, return the cached image
+      const cachedResizeCid = await getCachedResize(cid, height, width)
+      if (cachedResizeCid)
+        return Readable.from(
+          ipfs.cat(cachedResizeCid, { timeout: IPFS_TIMEOUT })
+        )
+
+      // If not, resize the image
+      const original = await getDataFromIPFS(cid)
+      const resizeOptions: sharp.ResizeOptions = {
+        fit: 'cover',
+        withoutEnlargement: true,
+      }
+      if (height) resizeOptions.height = height
```

Comment:
Didn't we check for `height` and `width` above?

Reply:
resizeOptions doesn't take `undefined`, so I didn't want to add it to the object if it's undefined.